src/libraries/System.IO.Hashing/src/System/IO/Hashing/XxHash3.cs

```diff
@@ -0,0 +1,884 @@
+        }
+
+        private static ulong HashLength129To240(byte* source, uint length, ulong seed)
+        {
+            Debug.Assert(length >= 129 && length <= 240);
+
+            ulong accumulators = length * XxHash64.Prime64_1;
+
+            accumulators += Mix16Bytes(source + (16 * 0), 0xBE4BA423396CFEB8, 0x1CAD21F72C81017C, seed); // DefaultSecretUInt64[0], DefaultSecretUInt64[1]
+            accumulators += Mix16Bytes(source + (16 * 1), 0xDB979083E96DD4DE, 0x1F67B3B7A4A44072, seed); // DefaultSecretUInt64[2], DefaultSecretUInt64[3]
+            accumulators += Mix16Bytes(source + (16 * 2), 0x78E5C0CC4EE679CB, 0x2172FFCC7DD05A82, seed); // DefaultSecretUInt64[4], DefaultSecretUInt64[5]
+            accumulators += Mix16Bytes(source + (16 * 3), 0x8E2443F7744608B8, 0x4C263A81E69035E0, seed); // DefaultSecretUInt64[6], DefaultSecretUInt64[7]
+            accumulators += Mix16Bytes(source + (16 * 4), 0xCB00C391BB52283C, 0xA32E531B8B65D088, seed); // DefaultSecretUInt64[8], DefaultSecretUInt64[9]
+            accumulators += Mix16Bytes(source + (16 * 5), 0x4EF90DA297486471, 0xD8ACDEA946EF1938, seed); // DefaultSecretUInt64[10], DefaultSecretUInt64[11]
+            accumulators += Mix16Bytes(source + (16 * 6), 0x3F349CE33F76FAA8, 0x1D4F0BC7C7BBDCF9, seed); // DefaultSecretUInt64[12], DefaultSecretUInt64[13]
+            accumulators += Mix16Bytes(source + (16 * 7), 0x3159B4CD4BE0518A, 0x647378D9C97E9FC8, seed); // DefaultSecretUInt64[14], DefaultSecretUInt64[15]
+
+            accumulators = Avalanche(accumulators);
+
+            if (length >= 16 * 9)
+            {
+                accumulators += Mix16Bytes(source + (16 * 8), 0x81017CBE4BA42339, 0x6DD4DE1CAD21F72C, seed); // DefaultSecret[3], DefaultSecret[11]
+                if (length >= 16 * 10)
+                {
+                    accumulators += Mix16Bytes(source + (16 * 9), 0xA44072DB979083E9, 0xE679CB1F67B3B7A4, seed); // DefaultSecret[19], DefaultSecret[27]
+                    if (length >= 16 * 11)
+                    {
+                        accumulators += Mix16Bytes(source + (16 * 10), 0xD05A8278E5C0CC4E, 0x4608B82172FFCC7D, seed); // DefaultSecret[35], DefaultSecret[43]
+                        if (length >= 16 * 12)
+                        {
+                            accumulators += Mix16Bytes(source + (16 * 11), 0x9035E08E2443F774, 0x52283C4C263A81E6, seed); // DefaultSecret[51], DefaultSecret[59]
+                            if (length >= 16 * 13)
+                            {
+                                accumulators += Mix16Bytes(source + (16 * 12), 0x65D088CB00C391BB, 0x486471A32E531B8B, seed); // DefaultSecret[67], DefaultSecret[75]
+                                if (length >= 16 * 14)
+                                {
+                                    accumulators += Mix16Bytes(source + (16 * 13), 0xEF19384EF90DA297, 0x76FAA8D8ACDEA946, seed); // DefaultSecret[83], DefaultSecret[91]
+                                    if (length >= 16 * 15)
```

Comment:
Since you're already unrolling loops and manually inlining constants, I feel like I should ask: Would it be better to make this use a flattend jumpable? e.g.

```C#
switch ((length - 16 * 8) / 16)
{
    case 0:
        break;
    case 1:
        accumulators += Mix16Bytes(source + (16 * 8), 0x81017CBE4BA42339, 0x6DD4DE1CAD21F72C, seed);
        break;
    case 2:
        accumulators += Mix16Bytes(source + (16 * 8), 0x81017CBE4BA42339, 0x6DD4DE1CAD21F72C, seed);
        accumulators += Mix16Bytes(source + (16 * 9), 0xA44072DB979083E9, 0xE679CB1F67B3B7A4, seed);
        break;
     ...
}
```

It might have too much unrolling cost.

The ideal assembly output, I feel, would be to have the 16*14 followed by 16*13, ..., and it jumps past all the ones it can't do.  But I don't know a C# pattern to get it to do that.

Since all that happens is += they can be reordered between the first and second calls to avalanche.

Reply:
[Sharplab suggests](https://sharplab.io/#v2:C4LghgzgtgNAJiA1AHwAICYCMBYAUKgZgAIMiBhIgbzyNpOIEsA7YIgWQAoAlAUzDgDyTADYBPAMoAHMEwA8AI1HAeAPiJwwwMDCLNWwnkwDmwABYBKGnWq46d3QDMiHA8bNE1GrQDoAMoZMLK3t7MwAnAHsAdyImHhiAQTCjAFcoQ2AAUQAPAGMeSWAGCKYOJjB0iIcXALNzcwBuYPtmuz0iMFzcogBeIgBadCbbEKJWuggohmBc02cOFL1zLzA/WqCRkJtR+zgeBzAU4VBxndpO7sQ+lYBtAA4AXWGz0aMI4AiiXMgeIjvnl5fH5/ECnM4XIhXdSaMA3ADsTzBOzeHyBEF+cIBL2+6KIcNBm0B5y6kOuMJuADZEYSiSjPjjfhSsWcGUQKQSiXYIVDbgBWamcuh0tG/XnMnas3kcwUdEk88kAFgFMuFrIV4tGaulgu5ZK0NwIysFquBBA1IVZBG1nN10P16CNnJNuKGSLorPQ1qJttumEdtPe9OBmHN9lZmC9gJ95IADP7Ac7fjHQ3ZWTHIy95GE+ABrUMAXzBYNQcNluWZhdw+aAA=) that a switch+goto-case might do it:

```C#
if (length >= 16 * 9)
{
    switch (the correct math)
    {
        default: // case 7
            accumulators += Mix16Bytes(source + (16 * 14), 0xBBDCF93F349CE33F, 0xE0518A1D4F0BC7C7, seed); // DefaultSecret[99], DefaultSecret[107]
            goto case 6;
        case 6:
            accumulators += Mix16Bytes(source + (16 * 13), 0xEF19384EF90DA297, 0x76FAA8D8ACDEA946, seed); // DefaultSecret[83], DefaultSecret[91]
            goto case 5;
        ...
        case 0:
            // maybe with this you don't need the if?
            break;
   }
}
```

Reply:
Good idea!  I'd convinced myself I couldn't do that because of ordering dependencies as part of computing the hash, but in this section there aren't such dependencies.